src/pybind/mgr/devicehealth/module.py

```diff
@@ -378,7 +378,7 @@ def do_scrape_daemon(self, daemon_type, daemon_id, devid=''):
         r, outb, outs = result.wait()
 
         try:
-            return json.loads(outb)
+            return json.loads(outb or outs)
```

Comment:
In  `src/pybind/mgr/devicehealth/module.py(381)do_scrape_daemon()` of master for an NVMe. I don't know if it's the same for HDDs.

```
user@ceph-1 /ceph (master*) $ while ! rlwrap nc localhost 5555 ; do sleep 1 ; done
> /ceph/src/pybind/mgr/devicehealth/module.py(381)do_scrape_daemon()
-> try:
(Pdb) l
376                 'devid': devid,
377             }), '')
378             r, outb, outs = result.wait()
379             import rpdb; rpdb.set_trace('0.0.0.0', 5555)
380  
381  ->         try:
382                 return json.loads(outb)
383             except (IndexError, ValueError):
384                 self.log.error(
385                     "Fail to parse JSON result from daemon {0}.{1} ({2})".format(
386                         daemon_type, daemon_id, outb))
(Pdb) pp r
0
(Pdb) pp outb
''
(Pdb) pp outs
('{\n'
 '    "Samsung_SSD_970_EVO_Plus_1TB_S4XXXXXXXX" : {\n'
 '        "device" : {\n'
 '            "info_name" : "/dev/nvme0n1",\n'
 '            "name" : "/dev/nvme0n1",\n'
 '            "protocol" : "NVMe",\n'
[...]
```

Reply:
In  `src/pybind/mgr/devicehealth/module.py(381)do_scrape_daemon()` of master for an NVMe. I don't know if it's the same for HDDs.

```
user@ceph-1 /ceph (master*) $ while ! rlwrap nc localhost 5555 ; do sleep 1 ; done
> /ceph/src/pybind/mgr/devicehealth/module.py(381)do_scrape_daemon()
-> try:
(Pdb) l
376                 'devid': devid,
377             }), '')
378             r, outb, outs = result.wait()
379             import rpdb; rpdb.set_trace('0.0.0.0', 5555)
380  
381  ->         try:
382                 return json.loads(outb)
383             except (IndexError, ValueError):
384                 self.log.error(
385                     "Fail to parse JSON result from daemon {0}.{1} ({2})".format(
386                         daemon_type, daemon_id, outb))
(Pdb) pp r
0
(Pdb) pp outb
''
(Pdb) pp outs
('{\n'
 '    "Samsung_SSD_970_EVO_Plus_1TB_S4XXXXXXXX" : {\n'
 '        "device" : {\n'
 '            "info_name" : "/dev/nvme0n1",\n'
 '            "name" : "/dev/nvme0n1",\n'
 '            "protocol" : "NVMe",\n'
[...]
```

Reply:
@p-se could you rebuild "ceph-mon", and try again? in my case
```
bin/ceph tell mon.a smart > /tmp/stdout 2>/tmp/stderr
```
the output was sent to stdout.

Reply:
@tchaikov Without recompiling my environment, I get the same result with the same command as you do.

```
user@ceph-1 /ceph/build (master*) $ ./bin/ceph tell mon.a smart > /tmp/stdout 2>/tmp/stderr
user@ceph-1 /ceph/build (master*) $ cat /tmp/stderr 
user@ceph-1 /ceph/build (master*) $ cat /tmp/stdout | head
{
    "Samsung_SSD_970_EVO_Plus_1TB_S4EWXXXXXXX" : {
        "device" : {
            "info_name" : "/dev/nvme0n1",
            "name" : "/dev/nvme0n1",
            "protocol" : "NVMe",
            "type" : "nvme"
        },
        "firmware_version" : "1B2QEXM7",
        "json_format_version" : [
user@ceph-1 /ceph/build (master*) $ 

```

Reply:
@p-se thanks for testing. no need to do so. so it sounds like a bug on mgr side. but in my testbed i cannot reproduce this issue with
```
$ bin/ceph device query-daemon-health-metrics osd.0
```
this command is able to query the smart info of disk mounted on osd.0.

Reply:
In contrast to `ceph tell ...`, querying the data on the console using `ceph device query-daemon-health-metrics` returned it in stderr for me, so that should be the right command to test it. I wonder if it depends on the type of the drive. I will test that, too.

```
user@ceph-1 /ceph/build (master*) $ { ./bin/ceph device query-daemon-health-metrics osd.0 ;}  >/tmp/stdout 2>/tmp/stderr
user@ceph-1 /ceph/build (master*) $ cat /tmp/stdout        
user@ceph-1 /ceph/build (master*) $ head /tmp/stderr
{
    "Samsung_SSD_970_EVO_Plus_1TB_S4EXXXXXX" : {
        "device" : {
            "info_name" : "/dev/nvme0n1",
            "name" : "/dev/nvme0n1",
            "protocol" : "NVMe",
            "type" : "nvme"
        },
        "firmware_version" : "1B2QEXM7",
        "json_format_version" : [
user@ceph-1 /ceph/build (master*) $ 

```

Reply:
What version are you running, both on the server/daemon side and for the 'ceph' CLI tool?

Reply:
This is a vstart cluster on current master (latest commit is ae17f3678d2106074858a58abe6325ac1854d8b9). Due to that I assume server and daemon are all on the same version. I'm not sure about that though, so if you want me to check something particular please let me know.

```
user@ceph-1 /ceph (master*) $ ceph version
ceph version Development (no_version) octopus (dev)
```

Built with `-DENABLE_GIT_VERSION=OFF`, so no version in that field.

```
smartmontools release 7.0 dated 2018-12-30 at 14:47:55 UTC
smartmontools SVN rev 4917 dated 2019-05-21 at 20:14:34
```

Reply:
Hmm, I rechecked locally on latest master and it seems to be behaving for me:
```
gnit:build ((13d0a2417c0...)) 09:51 AM $ bin/ceph tell mon.a smart 2> e > o ; wc -l e o
   2 e
  73 o
  75 total
gnit:build ((13d0a2417c0...)) 09:52 AM $ git describe
v15.0.0-6760-g13d0a2417c0
```

Reply:
Using `ceph tell` might not reveal the issue. If I use `ceph tell`, it looks okay, but if I run `ceph device query-daemon-health-metrics` it returns in stderr.

```
user@home /mnt/usb/ceph/build ±master⚡ » bin/ceph tell mon.a smart 2> e > o ; wc -l e o
    0 e
  666 o
  666 total
user@home /mnt/usb/ceph/build ±master⚡ » { ./bin/ceph device query-daemon-health-metrics osd.0 ;}  2>e >o ; wc -l e o        
  666 e
    0 o
  666 total
```

Btw this is on a vstart cluster on a regular HDD.

Reply:
weird,
```
gnit:build (wip-version-no-v) 07:18 AM $ bin/ceph tell mon.a smart 2> e > o ; wc -l e o
   0 e
  73 o
  73 total
gnit:build (wip-version-no-v) 07:18 AM $ bin/ceph device query-daemon-health-metrics mon.a 2> e > o ; wc -l e o
   0 e
  73 o
  73 total
```

Reply:
aha... i can reproduce it on an osd, but not a mon!

Reply:
see https://github.com/ceph/ceph/pull/31412